ENIGMAsystem/SHELL/Graphics_Systems/OpenGL1/GLvertex.cpp

```diff
@@ -123,31 +156,31 @@ ClientState graphics_apply_vertex_format(int format) {
       case vertex_type_ubyte4: elements = 4; size = 1; type = GL_UNSIGNED_BYTE; break;
     }
 
-    #define GL_ATTRIB_OFFSET(P) ((const GLvoid *) (sizeof(gs_scalar) * P))
+    #define GL_ATTRIB_OFFSET(P) ((const GLvoid *) ((intptr_t)base_pointer + (sizeof(gs_scalar) * P)))
 
     // this is an "emulation" of vertex format declarations for the OpenGL fixed-function pipeline
     switch (flag.second) {
       case vertex_usage_position:
-        if (state.useVertices) break;
-        state.useVertices = true;
+        if (useVertices) break;
```

Comment:
I didn't think calling glGet each iteration of the loop would be efficient. Also, it's not a matter of me wanting to do this, it's a requirement. I tested what GMS does, using its default shader, when you specify two colors in a vertex format. It only uses the first color that appears for drawing. So if I remove the bool check in my code, ENIGMA would use the last color, and we don't want it doing that. Because GL1 has no equivalent to `IDirect3D9VertexDeclaration` or a VAO, I could maybe have it cache a "flattened" copy of the vertex format that would be quicker to apply. That's unlikely to change our numbers on the cubes demo however.

Reply:
This sucks a bunch of ass: 740a263ba495a89e8c3ed064496bc5e4e640932d

Allow me to explain. `glPushClientAttrib` and `glPopClientAttrib` are really slow (totally killed the framerate), just like I suggested `glGet` would be, they involve a round trip to the GPU. Hence, you have to cache it yourself. I had to revert back to my even more verbose manual client state caching to speed back up my tiles benchmark that I didn't check until just now.

I looked high and low for an alternative here, and there's just no way we can avoid the Enable/Disable with our own caching, this is just how people do it. It allows other render functions (e.g, specifically `d3d_draw_floor` in Wild Racing GMK) to safely draw with vertex arrays which are faster than and just as expressive as immediate mode. It's just not good to leave certain client state enabled if we also want to allow user extensions to submit OpenGL calls to our context.

Now, I may have a way to avoid all of this in a future pr but it requires a lot of discussing. I am talking about moving `GLprimitives.cpp` with `draw_primitive_begin` and such over to `General/GSprimitives.cpp` for all of the systems. The functions would use the model class to define primitives and the model would be actually drawn and cleared (aka "flushed") whenever a render state change occurs (we already do this in GL3/DX9). Then we could totally cache the enabled client states since we won't have to worry about any other part of the system using vertex arrays. However, that would again be iffy if we want to allow extensions to submit GL calls to our context, but they could totally just "ask" us what the current vertex format is through API.